### PR TITLE
remote-ssh: fix Node not patching automatically

### LINKED ISSUE #103: Failed to connect to the remote extension host server due to unsupported, out of date libraries like glibc (e.g. older CentOS version)
- VSCode Version: Version 1.34.0-insider (1.34.0-insider)
- Local OS Version: macOS 10.14.4 (18E226)
- Remote OS Version: Amazon Linux Bare Metal release 2012.03
- Remote Extension/Connection Type: SSH

Steps to Reproduce:

1. Cmd-Shift-P, and choose *Remote-SSH: Connect current window to host...*
2. Connection is successful. Remote tools are installed and server is running on remote
3. VS Code shows an error at right-bottom: `Failed to connect to the remote extension host server (Error: Connection error: Unauthorized client refused.)`

`AllowTcpForwarding` is enabled on the remote server.

Here is the full output from "Remote - SSH" channel:

```
SSH Resolver called for "ssh-remote+dev-server"
SSH Resolver called for host: dev-server
Setting up SSH remote "dev-server"
Using commit id "473af338e1bd9ad4d9853933da1cd9d5d9e07dc9" and quality "insider" for server
Install and start server if needed
> bash: no job control in this shell
> Found existing installation...
> Warning: bad syntax, perhaps a bogus '-'? See /usr/share/doc/procps-3.2.8/FAQ
> Found running server...
>  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
>  
> bb80525e-2122-4fa5-8e4a-9d91821eeb83==45539==
"install" terminal command done
Received install output: bb80525e-2122-4fa5-8e4a-9d91821eeb83==45539==
Server is listening on port 45539
Spawning tunnel with: ssh  dev-server -N -L localhost:32546:localhost:45539
Spawned SSH tunnel between local port 32546 and remote port 45539
Waiting for ssh tunnel to be ready
Tunneling remote port 45539 to local port 32546
Resolving "ssh-remote+dev-server" to "localhost:32546", attempt: 1
```

Logs from "Log (Window)" channel:
```
[2019-05-03 18:02:25.843] [renderer1] [error] Connection error: Unauthorized client refused.: Error: Connection error: Unauthorized client refused.
    at v (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1544:9)
    at t.onControlMessage.o (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1542:930)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at a (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:48)
    at e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:94)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at S._receiveMessage (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:330:65)
    at S._socketDisposables.push._socketReader.onMessage.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:327:243)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at g.acceptChunk (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:801)
    at _register._socket.onData.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:153)
    at Socket.t (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:332:235)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:636:20)
[2019-05-03 18:02:25.844] [renderer1] [error] Cannot read property 'extensions' of null: TypeError: Cannot read property 'extensions' of null
    at e.<anonymous> (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:3850:751)
    at Generator.next (<anonymous>)
    at r (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:34:454)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[2019-05-03 18:02:25.844] [renderer1] [error] Connection error: Unauthorized client refused.: Error: Connection error: Unauthorized client refused.
    at v (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1544:9)
    at t.onControlMessage.o (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1542:930)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at a (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:48)
    at e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:94)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at S._receiveMessage (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:330:65)
    at S._socketDisposables.push._socketReader.onMessage.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:327:243)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at g.acceptChunk (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:801)
    at _register._socket.onData.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:153)
    at Socket.t (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:332:235)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:636:20)
[2019-05-03 18:02:28.207] [renderer1] [error] Could not fetch environment: Error: Could not fetch environment
    at d.getEnvironment.then.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:3813:413)
```

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by Enzime**
###### Description of changes

The code in #76202 has since stopped working due to `extension.js` changing. I've changed the location of the code insertion to ensure it runs every time before attempting to start the remote server.

The code now tests if the VS Code Server's Node is working and the correct version before attempting to patch it. It will first attempt to build Node from Nixpkgs, if that fails, then it will fallback to the locally installed Node.

Fixes https://github.com/microsoft/vscode-remote-release/issues/103 and https://github.com/microsoft/vscode-remote-release/issues/648

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 1]
```diff
@@ -9,29 +9,61 @@
...
-        echo PATCH: replacing ''$f with ''$localNodePath
-        rm ''$f
-        ln -s ''$localNodePath ''$f
+    echo "VS Code Node: $f"
+
+    # Check if VS Code Server has a non-working Node or the wrong version of Node
+    if ! nodeVersion=$($f -v) || [ "\''${nodeVersion:1:2}" != "14" ]; then
+      echo "VS Code Node Version: $nodeVersion"
+
+      nodePath=$(nix-build "<nixpkgs>" -A nodejs-14_x --no-out-link)/bin/node
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix:23]: Won't that be garbage collected?

Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix:23]: the code runs on every connect, so if it stops working for whatever reason (like Node being gc'd) it'll just build Node again

Comment (SuperSandro2000) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix:23]: Could we easily prevent it from being garbage collected like creating an out link?

Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: Fixed :+1:

**Comment by Enzime** [comment]
Added tests to ensure that Remote SSH works :smile:

**Comment by Enzime** [comment]
@ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
@ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
https://github.com/ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
@ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
@ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
https://github.com/ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
@ofborg test vscode-remote-ssh

**Comment by Enzime** [comment]
@ofborg build vscode

**Comment by SuperSandro2000** [comment]
You can edit old comments and add a space to retrigger ofborg.

**Comment by Enzime** [comment]
@ofborg eval

**Pull request closed without merging**